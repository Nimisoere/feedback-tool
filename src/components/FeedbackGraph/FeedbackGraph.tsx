import React from 'react'
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
} from 'chart.js';
import { Bar } from 'react-chartjs-2';
import useGetFeedBack from '../../hooks/api/useGetFeedBack';
import { generateChartObject } from './graph.utils';
import Skeleton from '../common/Skeleton';

ChartJS.register(
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
);

export const options = {
  responsive: true,
  plugins: {
    title: {
      display: true,
      text: 'Feedback Distribution',
    },
  },
};

type Props = {}

const FeedbackGraph = (props: Props) => {
  const { feedbacks, error, isLoading, getFeedback } = useGetFeedBack();
  if (error) {
    return (
      <div role="status" className="w-full">
        <div className='alert-danger'>
          <span className='font-medium'>{error?.message || "Could not fetch feedback list at this time"}</span>
          <button type='button' onClick={() => getFeedback()} className='btn-danger'>
            Try again
          </button>
        </div>
      </div>
    )
  }

  if (isLoading) {
    return <Skeleton />
  }

  const data = generateChartObject(feedbacks || [])

  const chartData = {
    labels: data.map(row => row.rating),
    datasets: [
      {
        data: data.map(row => row.count),
      }
    ]
  }

  return (
    <Bar options={options} data={chartData} />
  )
}

export default FeedbackGraph