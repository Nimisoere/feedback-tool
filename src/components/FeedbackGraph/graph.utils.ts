import { FeedbackResponse } from "../../hooks/api/useGetFeedBack";
import { groupBy } from "../../utils/common";

type FeedbackChartData = {
  rating: string;
  count: number;
}

export const generateChartObject = (data: FeedbackResponse[]):FeedbackChartData[] => {
  const groupedFeedbackByRatings = groupBy(data, "rating");
  return Object.keys(groupedFeedbackByRatings)?.map((rating) => ({
    count: groupedFeedbackByRatings[rating]?.length,
    rating: `${rating} Star(s)`
  }))
}