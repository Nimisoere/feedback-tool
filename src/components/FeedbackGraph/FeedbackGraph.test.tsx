import React from "react";
import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import FeedbackGraph from "./FeedbackGraph";

describe("Feedback Graph", () => {
  it("should match snapshot", async () => {
    const { getByText, container } = render(
      <FeedbackGraph />
    );
    expect(container).toMatchSnapshot();
  });
});
