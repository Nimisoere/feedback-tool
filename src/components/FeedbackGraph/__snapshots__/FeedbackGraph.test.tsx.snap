// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback Graph should match snapshot 1`] = `
<div>
  <div
    class="animate-pulse"
    role="status"
  >
    <div
      class="h-2.5 bg-gray-200 rounded-full max-w-[250px] mb-4"
    />
    <div
      class="h-2 bg-gray-200 rounded-full max-w-[360px] mb-2.5"
    />
    <div
      class="h-2 bg-gray-200 rounded-full mb-2.5"
    />
    <div
      class="h-2 bg-gray-200 rounded-full max-w-[330px] mb-2.5"
    />
    <div
      class="h-2 bg-gray-200 rounded-full max-w-[300px] mb-2.5"
    />
    <div
      class="h-2 bg-gray-200 rounded-full max-w-[360px]"
    />
    <span
      class="sr-only"
    >
      Loading...
    </span>
  </div>
</div>
`;
